### DISCUSSION TITLE
Exploring Private Space as a Work Profile Alternative

### DISCUSSION THREAD

**Discussion by filipfalcon** [ImranR98/Obtainium#1931] [Q&A] [answered]
I'm exploring the potential of Android 15's [Private Space](https://source.android.com/docs/security/features/private-space) feature as a replacement for the traditional [Work Profile](https://www.android.com/enterprise/work-profile/) setup. One of the key benefits of Private Space is that it doesn't require a third-party app for setup.

I've been using [Shelter](https://gitea.angry.im/PeterCxy/Shelter) to manage my Work Profile, which allows me to clone apps into the profile. A useful side effect of this setup is that when I update an app through Obtainium on my main profile, the cloned version in the Work Profile also receives the update, since they share the same ID.

However, I'm having trouble replicating this behavior with Private Space. When I try to install Obtainium in the Private Space, I'm [blocked by the system due to the duplicate ID](https://discuss.grapheneos.org/d/16670-private-space-on-android-15-grapheneos/16). Ideally, I'd like to be able to clone the application installed by Obtainium on the main profile directly to the Private Space.

Can anyone suggest a workaround or approach to achieve this behavior with Private Space? I'd love to hear from others who have experimented with this feature.

**Comment by ImranR98** [comment]
Are you installing the exact same APK in the private space? The issue you linked to talks about apps that have the same ID but different signing keys.

**Reply by ImranR98** [reply]
I just tried installing Obtainium in the private space and had no issues.

**Reply by filipfalcon** [reply]
I'm trying to install the APK of the latest version: https://github.com/ImranR98/Obtainium/releases/download/v1.1.28/app-release.apk

It's not the same APK file that I used to install Obtainium on the main profile since it was a way more versions back. However, the Obtainium on the main profile is updated to 1.1.28.

**Reply by ImranR98** [reply]
Hmm, not sure what's going on here.
Could you use [AppVerifier](https://github.com/soupslurpr/AppVerifier/releases) to check the signing key of the Obtainium install in your main profile? Then do the same for the new APK you downloaded. They should match.

**Reply by filipfalcon** [reply]
I don't understand the difference between the APKs that are provided in the assets with each release. I've been trying to install Obtainium through `app-release.apk` since that sounded the most straightforward to me, but it failed. A moment ago I've tried to install it through `app-arm64-v8a-release.apk` and it installed successfully. Now I have Obtainium on the main profile and in the Private Space.

**Reply by ImranR98** [reply]
Then it may be a different issue from what you linked. Both of those APKs are signed with the same key. The difference is that `app-release` will work on any CPU architecture, while the other one will only work on armv8 (pretty much all phones released in the past several years). Maybe the app installed in private space has to be exactly identical to the one in main.

**Reply by filipfalcon** [reply]
Thank you for the assistance. To summarize, Obtainium can be installed on the Private Space profile even if it is currently installed on the main profile. Downloading the correct APK from GitHub Releases should do just fine.

However, since it's isolated from the rest of the system I don't think updating it on the main profile will update it in the Private Space. That's not an issue Obtainium itself can solve as it would be actually against the point of the Private Space. Have I got it correctly?

**Reply by filipfalcon** [reply, marked as answer]
> Thank you for the assistance. To summarize, Obtainium can be installed on the Private Space profile even if it is currently installed on the main profile. Downloading the correct APK from GitHub Releases should do just fine.
> 
> However, since it's isolated from the rest of the system I don't think updating it on the main profile will update it in the Private Space. That's not an issue Obtainium itself can solve as it would be actually against the point of the Private Space. Have I got it correctly?

I've revisited the issue and had a chance to test the scenario I previously outlined.

It turns out that having Obtainium installed on both the main profile and the Private Space profile allows you to update the Private Space instance by updating it through Obtainium on the main profile. This is because the applications are installed system-wide under the same package name. Notably, installing Obtainium in the Private Space is still necessary, as it doesn't have the ability to clone apps to the Private Space like Shelter does with the Work profile.
